### PR TITLE
nix: add p4-fusion building

### PR DESCRIPTION
**Pull request by Strum355**
Introduces Nix package [derivations](https://github.com/justinwoo/nix-shorts/blob/master/posts/your-first-derivation.md) for building Salesforce's [p4-fusion](https://github.com/salesforce/p4-fusion) on macOS (including apple silicon) and linux. Pls grill it, hurl questions at me, request changes, w/e. 

## Supported targets

1. aarch64 macOS
2. x86_64 macOS
3. x86_64 linux (portable, **glibc**)
4. x86_64 linux (nix)

## How to build

`nix build .#p4-fusion --verbose` for a macOS binary (on a macOS system), `nix build .#p4-fusion-portable` for a **glibc** binary for most glibc-based Linux system (not NixOS)

## Additional notes

~~The build for aarch64-darwin compiles an x86_64 binary that can be run transparently via Rosetta. Aint that some magic~~ Newer helix-core API provide native aarch64-darwin version, so we're using that

The repo bundles a number of dependencies for building `libgit2` including (`http-parser`, `zlib`, `ntlmclient` and `pcre`). Where they exist, the nix-managed versions are used instead. We can probably choose to use the bundled versions instead if people feel that would be better.

~~LibreSSL is used instead of OpenSSL, due to OpenSSL 1.0.2 being required but not available in the nixpkgs repository due to it being 1) unsupported upstream and 2) considered insecure. There are complications around linking with OpenSSL 1.1.0 or 3.0.x, with LibreSSL being significantly easier to get working. Functionally should pose no issues, potentially be more secure, and minimizes the amount of patching required.~~ Not necessary after upgrading to newer helix-core API, which supports OpenSSL 1.1.x

<details>
<summary><code>ldd</code> and <code>otool -L</code> output</summary>

Running `ldd` on the the `p4-fusion-portable` binary in a basic ubuntu container shows no missing libraries, as everything else is statically linked:
```
$ docker run -v (pwd)/p4-fusion:/data/p4-fusion -it --rm  ubuntu
root@ab2683978d52:/# ldd /data/p4-fusion 
        linux-vdso.so.1 (0x00007ffce23e1000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fc783371000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fc78336c000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fc783285000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fc783265000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc78303d000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fc78337a000)
```

Running `otool -L` on the `p4-fusion` binary on macOS shows the following. The binaries reference two entries from the nix store, but there appears to be no issue running them. Im chalking this up to the super mysterious dylib cache that exists as of BigSur, and which we can fairly safely assume as a lowest-supported-version. [<insert "shipping a built-in linker cache and removing shared objects from the filesystem, to own the libs," meme here>](https://twitter.com/mycoliza/status/1275307131017650176)

```
$ otool -L ./result/bin/p4-fusion 
./result/bin/p4-fusion:
	/System/Library/Frameworks/CFNetwork.framework/Versions/A/CFNetwork (compatibility version 1.0.0, current version 811.5.4)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1349.8.0)
	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa (compatibility version 1.0.0, current version 22.0.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 57740.60.18)
	/nix/store/wk6crafpdldhpn0vlvm6gz31f93f6ds6-libcxxabi-13.0.1/lib/libc++abi.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/nix/store/ywr1w4j9rq8whashkvbv57k39icqq5rn-libcxx-13.0.1/lib/libc++.1.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
```

</details>

Big props to [Randy Eckenrode](https://github.com/NixOS/nixpkgs/issues/180771#issuecomment-1317785816) for some crucial help in getting this unblocked.

## Test plan

Built on x86_64-linux: ran on host and in ubuntu docker container
Built on aarch64-darwin: ran on build host and on Horstens

### CODE REVIEW

**Comment by ryanslade** [comment]
Awesome, thanks for doing this!

Do we have any docs on what's required to install `nix` itself?
Where do we run `nix build .#p4-fusion --verbose`? In the root of the sourcegraph repo?

File: p4-fusion.nix [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+  buildInputs = [
+    libcxx
+    libcxxabi
+    zlib.static
+    zlib.dev
+    http-parser-static
+    pcre-static
+    # openssl 1.0.2 is considered insecure, and p4-fusion won't compile against openssl >1.1 or higher
+    # so we're using libressl instead
```
Comment (mrnugget) [inline p4-fusion.nix:89]: Nice. Might be useful enough to let SF know?

File: flake.nix [hunk 2]
```diff
@@ -35,5 +35,28 @@
           checkFlags = [ ];
         });
       };
+
+      # recursiveUpdate is just for recursively merging sets
+      packages = nixpkgs.lib.recursiveUpdate
+        {
+          aarch64-darwin.p4-fusion = self.packages.x86_64-darwin.p4-fusion;
+          x86_64-linux.p4-fusion-portable = self.packages.x86_64-linux.p4-fusion.overrideAttrs (oldAttrs: {
+            # patch the ELF interpreter for non-nix(os) distros.
```
Comment (mrnugget) [inline flake.nix:44]: Now that's a stop-in-your-tracks kinda sentence.

**Review by mrnugget — APPROVED** [review]
Very nice! I obviously have 0 useful things to add here, since I have 0 clue of nix (yet?). But this is already very helpful, even if just as documentation on how to get it to compile. I'll try to setup nix on macOS as soon as I can find the time.

File: flake.nix (hunk 2 above)
Comment (Strum355) [reply flake.nix:44]: Its some really cool and interesting stuff that really takes you down the road of understanding more about what `ldd` shows you

**Comment by ryanslade** [comment]
Some notes:

I followed the instructions to install nix on MacOS as linked above. Ran the install script and just said yes to everything, it seems to have worked. 

Tried running 

```
nix build .#p4-fusion --verbose
```

and got this error:

```
error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override
```

Had to add a few more experimental features and now I get this:

```
nix build .#p4-fusion --verbose --extra-experimental-features nix-command --extra-experimental-features flakes
error: flake 'git+file:///Users/ryan/work/sourcegraph' does not provide attribute 'packages.aarch64-darwin.p4-fusion', 'legacyPackages.aarch64-darwin.p4-fusion' or 'p4-fusion'
```

File: flake.nix [hunk 3]
```diff
@@ -35,5 +35,28 @@
...
+                $out/bin/p4-fusion
+            '';
+          });
+        }
+        (
+          nixpkgs.lib.genAttrs [ "x86_64-linux" "aarch64-linux" "x86_64-darwin" ] (system:
+            let pkgs = import nixpkgs { inherit system; };
+            in
+            {
+              p4-fusion = pkgs.callPackage ./p4-fusion.nix { };
```
Comment (keegancsmith) [inline flake.nix:57]: drive-by should we make this derivation live somewhere other than the top-level dir? Maybe under dev/nix for now?

Comment (Strum355) [reply flake.nix:57]: Sounds reasonable actually, yea. Will do that

File: p4-fusion.nix [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    '';
+  }));
+  libiconv-static = (libiconv.override { enableStatic = true; enableShared = false; });
+  libressl-static = (libressl.override { buildShared = false; });
+  libssh2-libressl = ((makeStatic libssh2).override { openssl = libressl-static; });
+  pcre-static = (makeStatic pcre).dev;
+in
+clang13Stdenv.mkDerivation rec {
+  name = "p4-fusion";
+  version = "v1.12";
```
Comment (ryanslade) [inline p4-fusion.nix:44]: What else would need to change to build a different release of `p4-fusion`?

**Comment by ryanslade** [comment]
Update: 

After actually checking out the correct branch it built 🎉

File: p4-fusion.nix (hunk 4 above)
Comment (Strum355) [reply p4-fusion.nix:44]: The [revhash and shasum](https://github.com/sourcegraph/sourcegraph/pull/44588/files#diff-491547b4041844d2ab0f37e68ce699c12a53c86e9f8e82aed35a3a1891660c75R51-R52) if just upstream p4-fusion needs to be bumped, and url and shasum [here](https://github.com/sourcegraph/sourcegraph/pull/44588/files#diff-491547b4041844d2ab0f37e68ce699c12a53c86e9f8e82aed35a3a1891660c75R59-R60) and [here](https://github.com/sourcegraph/sourcegraph/pull/44588/files#diff-491547b4041844d2ab0f37e68ce699c12a53c86e9f8e82aed35a3a1891660c75R66-R67) if updating helix-core API. That is, if they dont make some changes to e.g. openssl version requirements that arent compatible.

Theres a command that will give you the shasums ahead of time to fill in that I'll document somewhere once Im at a computer, nix will error at build time with the correct one as well though.

File: p4-fusion.nix (hunk 1 above)
Comment (Strum355) [reply p4-fusion.nix:89]: I think the incompatibility comes from helix-core API, which afaik they dont control. Perforce is someone else, right? 🤔

Comment (Strum355) [reply p4-fusion.nix:89]: Im curious to try a newer version of helix-core API to see if that would resolve the issue actually, Im trying to be mindful of how many hours I sink into this though :laff: could do it full-time

**Comment by Strum355** [comment]
Removed aarch64-linux target, theres no published helix-core-api for that.

Later helix-core API versions seem to include [native aarch64-darwin](https://filehost.perforce.com/perforce/r22.2/) distributions, so Im going to timebox an hour or two today to see if it successfully links with p4-fusion. Ideally this (newer helix-core API) would also be built for a newer openssl version so we might be able to drop the current "hacks".

File: flake.nix (hunk 2 above)
Comment (Strum355) [reply flake.nix:44]: Yea I havent specifically targeted or looked into muslc yet. The primary focus was getting something repo-management could use locally. @mrnugget is it intended to run it in a local alpine container or whats the usecase-flow? (Probably shouldve asked earlier :grimacing:)

File: p4-fusion.nix (hunk 1 above)
Comment (varungandhi-src) [reply p4-fusion.nix:89]: I'm a little confused; this is mentioning that OpenSSL is not being used, but there are a bunch of OpenSSL related bits later in this file. So is it being used or not? Or is part of the system being built as if it will be linking to OpenSSL even though it will actually be linked to LibreSSL?

**Comment by ryanslade** [comment]
FYI @twarit-waikar

File: p4-fusion.nix (hunk 1 above)
Comment (Strum355) [reply p4-fusion.nix:89]: Im not sure which part specifically youre referring to. `openssl = libressl` parts are making dependencies build against libressl instead of openssl, and the #if patching below patches a check for either libressl or openssl to also eval to for libressl >=3.5

Comment (varungandhi-src) [reply p4-fusion.nix:89]: - The filename below has `openssl` hard-coded in it (I guess that's incorrect, but we don't want to change that?)
- The macro that is being modified has `OPENSSL_VERSION_NUMBER >= 0x10100000L` in it (can we just remove that if we're modifying the `#if` anyways? Or is that kept as is because it was copied from the linked PR?)
- There is a comment which says "because the world of openssl is an API versioning nightmare" -- but we're not using OpenSSL, so LibreSSL versioning is the one that should matter, right?

File: flake.nix (hunk 2 above)
Comment (mrnugget) [reply flake.nix:44]: For now the scope is that we need to be able to run this on our M1 macs. We've had to ask someone from Salesforce to compile binaries for us. That's not good and this here is a fix for that. We already have p4-fusion in containers and this here is not (at least not now) aimed to replace that.

File: p4-fusion.nix (hunk 1 above)
Comment (Strum355) [reply p4-fusion.nix:89]: - Sure but that doesnt mean its openssl specific. Upstream also includes libressl checks because people would want to link with libressl instead of openssl :shrug: https://github.com/libgit2/libgit2/blob/main/deps/ntlmclient/crypt_openssl.c#L47-L49
- Also not sure what exactly the problem youre getting at here. Users can choose to link to either libressl or openssl (the former being a fork of the latter), so they have "largely" similar API surfaces but may diverge and need specific handling, as is the case here (see above also)
- LibreSSL is a fork of OpenSSL, OpenSSL is a fairly widely known awkward library to link against (very specific on what version you link it with), LibreSSL also deprecates and subsequently removes things from its API surface across minor version bumps (Im assuming they dont use semver, else that would be bad to break API in a minor version bump)

**Comment by Strum355** [comment]
[`93c19a5` (#44588)](https://github.com/sourcegraph/sourcegraph/pull/44588/commits/93c19a53b5f3673400140ae273f37227b199d21b) uses a newer helix-core API, which is designed for OpenSSL 1.1.x, as well as distributing a version for aarch64-darwin. Building and running on both x86_64-linux and aarch64-darwin was successful, so thats COOL and keeps things a bit tidier

File: p4-fusion.nix (hunk 4 above)
Comment (Strum355) [reply p4-fusion.nix:44]: Additional one needs to be updated if bumping helix-core API with the latest commit, for aarch64-darwin specific distribution

File: p4-fusion.nix (hunk 1 above)
Comment (Strum355) [reply p4-fusion.nix]: @varungandhi-src @mrnugget  I was able to remove requiring libressl (and the patching) with a newer helix-core API. Its newer than what salesforce are using upstream, but I was able to clone a repo locally, so I think we should be fairly OK

**Comment by Strum355** [comment]
Building a fat binary (x64+arm64) for macOS seems to be fairly trivial (invoke `lipo` on them), and building for x64 from arm64 doesnt seem to require any additional work, so I could add a `p4-fusion-fatbin` target to the two darwin system targets 🤷

**Comment by ChronicallySerious** [comment]
This is great to see on my time off! :D

I would recommend using the bundled libraries with p4-fusion, just because the fact that we are building those dependencies with specific build flags (https://github.com/salesforce/p4-fusion/blob/master/generate_cache.sh#L7) that reduce the binary size of the dependency, avoid all compiler incompatibility bugs when linking with code built by someone else (C++ is like that some times), and enable/disable new feature flags within those libraries once they are updated. 

It is basically an argument of having control over what code is being linked into the final binaries and how much of it can be avoided / optimized. I leave any further considerations up to you. 

Also, I have left half of this comment here - https://github.com/salesforce/p4-fusion/issues/28#issuecomment-1322570042

**Comment by Strum355** [comment]
> just because the fact that we are building those dependencies with specific build flags ... and enable/disable new feature flags within those libraries once they are updated.

Those flags (where relevant e.g. not the three cmake and openssl ones)  would have to be copied into the nix derivation either way, as we dont use `build.sh` or `generate_cache.sh`. We already copy some over (I see some need changing, like the SSH one), as seen here: https://github.com/sourcegraph/sourcegraph/pull/44588/files#diff-491547b4041844d2ab0f37e68ce699c12a53c86e9f8e82aed35a3a1891660c75R112-R123. The latter feature flags, if theyre passed the same way to CMake, would also need the same treatment. 

> that reduce the binary size of the dependency

Ironically, the binary I see you've sent on slack is 8.0MB, while the one built here is 7.0MB (6.8MB with ssh and https disabled), and this is without the bundled dependencies (besides direct dependencies libgit2 and minitrace, more details below) :grinning: not sure where the difference comes from, nix applies some stripping by default as a post-build step.

> avoid all compiler incompatibility bugs when linking with code built by someone else (C++ is like that some times)

C++ really do be like that sometimes (hence, and for other reasons, I created this nix derivation). We're still using the bundled libgit2 and minitrace, but using http-parser, libiconv (for macOS), pcre and zlib from nixpkgs (along with openssl of course). Where necessary , they'll be automatically built locally from source using gcc 11 on Linux and clang 11 on macOS (as the default nix builders for those platforms), same as what p4-fusion will get built with in this derivation. This hopefully addresses your concern here? (fixing up my PR on this note, as previously we had a mix of clang13+gcc11 on linux and clang13+clang11 on macOS).

I believe the approach we have here is also more optimal for nix to cache, as individual derivations are cached, we would have to make derivations per bundled dependency (to some level of coarseness) in order to reduce the amount of rebuilding that has to be done. We already have to recompile libgit2, minitrace and p4-fusion every time as theyre not separate derivations, while openssl n co only need rebuilding if the inputs change. Hope that makes sense as a quick primer on Nix caching :smile: 

Very much appreciate the input and concerns here, I definitely understand the arguments for using the bundled/vendored dependencies instead, and Im not super beholden to either approach. The combined compile times should be relatively reasonable Id imagine anyways. @ChronicallySerious

**Comment by ChronicallySerious** [comment]
I think that makes sense. I wasn't aware there are multiple binaries available per nix library, with caching in place where possible. 

The 8 MB binary sent via Slack is still a Debug mode binary so it may not be the best candidate to compare here. 

Binary stripping while linking would help reduce the size anyway but my point was more about the features that have been turned off explicitly in ligbit2 and the difference in the optimization levels across binaries that are brought from the cache and linked into the final program. 

However both of these is a moot point now that I realize that nix will not bring over cached binaries that don't match the compilation flags provided here. I believe that addresses everything I said above in favour of nix.

**Comment by Strum355** [comment]
Greenlit to be merged as is, docs and improvements/additions will come in follow up PRs. Thanks to everyone :comfy:
